### PR TITLE
Make more things async

### PR DESCRIPTION
**Pull request by davidfowl**
This is preparation for making it possible for resources to wait for a longer time for things like connection strings to be resolved. e.g. in the case of azure provisioning.

- Environment callbacks
- Added GetConnectionStringAsync
- Updated all of the tests
 ###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2533)

### CODE REVIEW

**Review by mitchdenny — APPROVED** [review]

File: playground/CustomResources/CustomResources.AppHost/TestResource.cs [hunk 1]
```diff
@@ -44,7 +44,7 @@ public Task BeforeStartAsync(DistributedApplicationModel appModel, CancellationT
                 Task.Run(async () =>
                 {
                     // Simulate custom resource state changes
-                    var state = resourceUpdates.GetInitialSnapshot();
+                    var state = await resourceUpdates.GetInitialSnapshotAsync();
```
Comment (JamesNK) [inline playground/CustomResources/CustomResources.AppHost/TestResource.cs:47]: Pass `_tokenSource.Token`

File: src/Aspire.Hosting/ApplicationModel/CommandLineArgsCallbackAnnotation.cs [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    /// </summary>
+    public Func<CommandLineArgsCallbackContext, Task> Callback { get; }
+}
+
+/// <summary>
+/// Represents a callback context for the list of command-line arguments associated with an executable resource.
+/// </summary>
+/// <param name="args"> The list of command-line arguments.</param>
+/// <param name="cancellationToken"> The cancellation token associated with this execution.</param>
+public class CommandLineArgsCallbackContext(IList<string> args, CancellationToken cancellationToken = default)
```
Comment (JamesNK) [inline src/Aspire.Hosting/ApplicationModel/CommandLineArgsCallbackAnnotation.cs:44]:
```suggestion
public sealed class CommandLineArgsCallbackContext(IList<string> args, CancellationToken cancellationToken = default)
```

File: src/Aspire.Hosting/ApplicationModel/CustomResourceExtensions.cs [hunk 3]
```diff
@@ -17,14 +17,27 @@ public static class CustomResourceExtensions
     /// <param name="builder">The resource builder.</param>
     /// <param name="initialSnapshotFactory">The factory to create the initial <see cref="CustomResourceSnapshot"/> for this resource.</param>
     /// <returns>The resource builder.</returns>
-    public static IResourceBuilder<TResource> WithResourceUpdates<TResource>(this IResourceBuilder<TResource> builder, Func<CustomResourceSnapshot>? initialSnapshotFactory = null)
+    public static IResourceBuilder<TResource> WithResourceUpdates<TResource>(this IResourceBuilder<TResource> builder, Func<ValueTask<CustomResourceSnapshot>>? initialSnapshotFactory = null)
```
Comment (JamesNK) [inline src/Aspire.Hosting/ApplicationModel/CustomResourceExtensions.cs:20]: Func should take a cancellation token?

File: src/Aspire.Hosting/ApplicationModel/ResourceUpdatesAnnotation.cs [hunk 4]
```diff
@@ -10,7 +10,7 @@ namespace Aspire.Hosting.ApplicationModel;
 /// <summary>
 /// The annotation that allows publishing and subscribing to changes in the state of a resource.
 /// </summary>
-public sealed class ResourceUpdatesAnnotation(Func<CustomResourceSnapshot> initialSnapshotFactory) : IResourceAnnotation
+public sealed class ResourceUpdatesAnnotation(Func<ValueTask<CustomResourceSnapshot>> initialSnapshotFactory) : IResourceAnnotation
```
Comment (JamesNK) [inline src/Aspire.Hosting/ApplicationModel/ResourceUpdatesAnnotation.cs:13]: Cancellation token

File: src/Aspire.Hosting/ApplicationModel/ResourceUpdatesAnnotation.cs [hunk 5]
```diff
@@ -24,7 +24,7 @@ public sealed class ResourceUpdatesAnnotation(Func<CustomResourceSnapshot> initi
     /// <summary>
     /// Gets the initial snapshot of the dashboard state for this resource.
     /// </summary>
-    public CustomResourceSnapshot GetInitialSnapshot() => initialSnapshotFactory();
+    public ValueTask<CustomResourceSnapshot> GetInitialSnapshotAsync() => initialSnapshotFactory();
```
Comment (JamesNK) [inline src/Aspire.Hosting/ApplicationModel/ResourceUpdatesAnnotation.cs:27]: Cancellation token

File: src/Aspire.Hosting/Dashboard/DcpDataSource.cs [hunk 6]
```diff
@@ -201,7 +201,7 @@ private async Task ProcessInitialResourceAsync(IResource resource, CancellationT
         {
             // We have a dashboard annotation, so we want to create a snapshot for the resource
             // and update data immediately. We also want to watch for changes to the dashboard state.
-            var state = resourceUpdates.GetInitialSnapshot();
+            var state = await resourceUpdates.GetInitialSnapshotAsync().ConfigureAwait(false);
```
Comment (JamesNK) [inline src/Aspire.Hosting/Dashboard/DcpDataSource.cs:204]: Cancellation token

File: src/Aspire.Hosting/Extensions/ExecutableResourceBuilderExtensions.cs [hunk 7]
```diff
@@ -61,14 +61,14 @@ public static IResourceBuilder<ExecutableResource> AddExecutable(this IDistribut
...
-        var annotation = new ExecutableArgsCallbackAnnotation(updatedArgs =>
+        var annotation = new CommandLineArgsCallbackAnnotation(updatedArgs =>
         {
             updatedArgs.AddRange(args);
         });
         return builder.WithAnnotation(annotation);
     }
 
-    private static void WriteExecutableAsDockerfileResource(ManifestPublishingContext context, ExecutableResource executable)
+    private static async Task WriteExecutableAsDockerfileResource(ManifestPublishingContext context, ExecutableResource executable)
```
Comment (JamesNK) [inline src/Aspire.Hosting/Extensions/ExecutableResourceBuilderExtensions.cs:71]: nit: WriteExecutableAsDockerfileResource -> WriteExecutableAsDockerfileResourceAsync

File: src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs [hunk 8]
```diff
@@ -149,18 +158,18 @@ public void WriteBindings(IResource resource, bool emitContainerPort = false)
     /// TODO: Doc Comments
     /// </summary>
     /// <param name="resource"></param>
-    public void WriteEnvironmentVariables(IResource resource)
+    public async Task WriteEnvironmentVariables(IResource resource)
```
Comment (JamesNK) [inline src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs:161]: WriteEnvironmentVariablesAsync

Cancellation token

File: src/Aspire.Hosting/MySql/MySqlServerResource.cs [hunk 9]
```diff
@@ -53,9 +53,9 @@ internal void AddDatabase(string name, string databaseName)
         _databases.TryAdd(name, databaseName);
     }
 
-    internal void WriteToManifest(ManifestPublishingContext context)
+    internal async Task WriteToManifest(ManifestPublishingContext context)
     {
-        context.WriteContainer(this);
+        await context.WriteContainer(this).ConfigureAwait(false);
```
Comment (JamesNK) [inline src/Aspire.Hosting/MySql/MySqlServerResource.cs:58]: WriteContainerAsync. Cancellation token? (or is it on the the context?)

File: src/Aspire.Hosting/Oracle/OracleDatabaseServerResource.cs [hunk 10]
```diff
@@ -53,9 +53,9 @@ internal void AddDatabase(string name, string databaseName)
         _databases.TryAdd(name, databaseName);
     }
 
-    internal void WriteToManifest(ManifestPublishingContext context)
+    internal async Task WriteToManifest(ManifestPublishingContext context)
```
Comment (JamesNK) [inline src/Aspire.Hosting/Oracle/OracleDatabaseServerResource.cs:56]: WriteToManifestAsync. CT?

File: src/Aspire.Hosting/Postgres/PostgresServerResource.cs [hunk 11]
```diff
@@ -68,9 +83,9 @@ internal void AddDatabase(string name, string databaseName)
         _databases.TryAdd(name, databaseName);
     }
 
-    internal void WriteToManifest(ManifestPublishingContext context)
+    internal async Task WriteToManifest(ManifestPublishingContext context)
```
Comment (JamesNK) [inline src/Aspire.Hosting/Postgres/PostgresServerResource.cs:86]: nit: WriteToManifestAsync

File: src/Aspire.Hosting/Publishing/ManifestPublisher.cs [hunk 12]
```diff
@@ -48,68 +48,69 @@ protected virtual async Task PublishInternalAsync(DistributedApplicationModel mo
...
         jsonWriter.WriteStartObject();
-        WriteResources(model, context);
+        await WriteResources(model, context).ConfigureAwait(false);
         jsonWriter.WriteEndObject();
 
         await jsonWriter.FlushAsync(cancellationToken).ConfigureAwait(false);
     }
 
-    private static void WriteResources(DistributedApplicationModel model, ManifestPublishingContext context)
+    private static async Task WriteResources(DistributedApplicationModel model, ManifestPublishingContext context)
```
Comment (JamesNK) [inline src/Aspire.Hosting/Publishing/ManifestPublisher.cs:60]: nit: Add Async suffix to methods where needed

File: src/Aspire.Hosting/Oracle/OracleDatabaseServerResource.cs (hunk 10 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Oracle/OracleDatabaseServerResource.cs:56]: It's on the context

File: src/Aspire.Hosting/MySql/MySqlServerResource.cs (hunk 9 above)
Comment (davidfowl) [reply src/Aspire.Hosting/MySql/MySqlServerResource.cs:58]: It's on the context

File: src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs (hunk 8 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Publishing/ManifestPublishingContext.cs:161]: It's on the context

File: src/Aspire.Hosting/Publishing/ManifestPublisher.cs (hunk 12 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Publishing/ManifestPublisher.cs]: Done

File: src/Aspire.Hosting/Postgres/PostgresServerResource.cs (hunk 11 above)
Comment (davidfowl) [reply src/Aspire.Hosting/Postgres/PostgresServerResource.cs]: Done

File: src/Aspire.Hosting/ApplicationModel/ResourceUpdatesAnnotation.cs (hunk 5 above)
Comment (davidfowl) [reply src/Aspire.Hosting/ApplicationModel/ResourceUpdatesAnnotation.cs]: Done

File: src/Aspire.Hosting/ApplicationModel/ResourceUpdatesAnnotation.cs (hunk 4 above)
Comment (davidfowl) [reply src/Aspire.Hosting/ApplicationModel/ResourceUpdatesAnnotation.cs]: Done

File: src/Aspire.Hosting/ApplicationModel/CustomResourceExtensions.cs (hunk 3 above)
Comment (davidfowl) [reply src/Aspire.Hosting/ApplicationModel/CustomResourceExtensions.cs]: Done

File: src/Aspire.Hosting/ApplicationModel/CommandLineArgsCallbackAnnotation.cs (hunk 2 above)
Comment (davidfowl) [reply src/Aspire.Hosting/ApplicationModel/CommandLineArgsCallbackAnnotation.cs]: Done

File: playground/CustomResources/CustomResources.AppHost/TestResource.cs (hunk 1 above)
Comment (davidfowl) [reply playground/CustomResources/CustomResources.AppHost/TestResource.cs]: Done

File: src/Aspire.Hosting/ApplicationModel/CustomResourceExtensions.cs [hunk 13]
```diff
@@ -17,14 +17,27 @@ public static class CustomResourceExtensions
     /// <param name="builder">The resource builder.</param>
     /// <param name="initialSnapshotFactory">The factory to create the initial <see cref="CustomResourceSnapshot"/> for this resource.</param>
     /// <returns>The resource builder.</returns>
-    public static IResourceBuilder<TResource> WithResourceUpdates<TResource>(this IResourceBuilder<TResource> builder, Func<CustomResourceSnapshot>? initialSnapshotFactory = null)
+    public static IResourceBuilder<TResource> WithResourceUpdates<TResource>(this IResourceBuilder<TResource> builder, Func<CancellationToken, ValueTask<CustomResourceSnapshot>>? initialSnapshotFactory = null)
```
Comment (JamesNK) [inline src/Aspire.Hosting/ApplicationModel/CustomResourceExtensions.cs:20]: Having a context here would reduce the chance of needing something extra in the future causing a breaking change. Up to you.

**Review by JamesNK — APPROVED** [review]

**Pull request merged**